### PR TITLE
[geometry/optimization] Enable preprocessing GCS before solving ShortestPath

### PR DESCRIPTION
**Pull request by mpetersen94**
Adds an optional preprocessing step to the `SolveShortestPath`.  Currently off by default to preserve the current behaviors and it on average doubles the time to run on simple test problems.

Depends on #17396 

cc @TobiaMarcucci @RussTedrake

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/17397)

### CODE REVIEW

**Review by xuchenhan-tri — COMMENTED** [review]
+(status: do not review) for now per platform review checklists.

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/17397#-:-N4ZauQM6bPRxAPoPyHx:bln1tm3)* status: needs platform reviewer assigned, needs at least two assigned reviewers, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits), missing label for release notes

**Review by mpetersen94 — COMMENTED** [review]
+@russtedrake for feature review
-(status: do not review) 

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/17397#-:-NAjoPgT6A8ySdVkzXFa:b-mi5c1e)* status: LGTM missing from assignee RussTedrake(platform), needs at least two assigned reviewers, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits), missing label for release notes (waiting on @RussTedrake)

**Review by RussTedrake — COMMENTED** [review]
I'm having trouble reviewing the two commits separately.  #17396 seems very close; maybe i'll review this as soon as that has landed and been rebased.

> Reviewed 1 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/17397#-:-NAk5FgmDxT9kteLk2bi:b-fm4e6p)* status: LGTM missing from assignee RussTedrake(platform), needs at least two assigned reviewers, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits), missing label for release notes (waiting on @RussTedrake)

**Review by RussTedrake — COMMENTED** [review]
seems like we need some test?  perhaps solve the same problem(s) with and without preprocessing and make sure we get the same answer?  but you can also make the testing class a friend and call the private preprocessing method directly.

> Reviewed 1 of 6 files at r2, 5 of 5 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/17397#-:-NAkSsJm-TL4mxTOT_0N:bkn1sk4)* status: 6 unresolved discussions, LGTM missing from assignee RussTedrake(platform), needs at least two assigned reviewers, missing label for release notes (waiting on @mpetersen94)

___
*[`geometry/optimization/graph_of_convex_sets.h` line 37 at r4](https://reviewable.io/reviews/robotlocomotion/drake/17397#-NAkSpKD7tuaf4bVFpkZ:-NAkSpKD7tuaf4bVFpk_:btfdvr9) ([raw file](https://github.com/robotlocomotion/drake/blob/b21d16d9bf564a6516237fdc31adb65ee6bea639/geometry/optimization/graph_of_convex_sets.h#L37)):*
> ```objc
>   path from source to target. In most cases, preprocessing causes a net
>   reduction in computation by reducing the size of the optimization solved. */
>   bool preprocessing{false};
> ```

i would think it should be true by default?
___
*[`geometry/optimization/graph_of_convex_sets.h` line 353 at r4](https://reviewable.io/reviews/robotlocomotion/drake/17397#-NAkSRsM5g8oVOnh-QVW:-NAkSRsM5g8oVOnh-QVX:bkeykf6) ([raw file](https://github.com/robotlocomotion/drake/blob/b21d16d9bf564a6516237fdc31adb65ee6bea639/geometry/optimization/graph_of_convex_sets.h#L353)):*
> ```objc
>   std::vector<const Edge*> Edges() const;
> 
>   /** Removes any constraints added to each edge with AddPhiConstraint. */
> ```

btw: perhaps "removes all constraints added to any edge with..?"
___
*[`geometry/optimization/graph_of_convex_sets.cc` line 266 at r4](https://reviewable.io/reviews/robotlocomotion/drake/17397#-NAkY0QkAlARnmMdunpD:-NAkY0QkAlARnmMdunpE:bbyfw05) ([raw file](https://github.com/robotlocomotion/drake/blob/b21d16d9bf564a6516237fdc31adb65ee6bea639/geometry/optimization/graph_of_convex_sets.cc#L266)):*
> ```cpp
> }
> 
> std::set<EdgeId> GraphOfConvexSets::PreprocessShortestPath(
> ```

can we perhaps cite Appendix A.2 in https://arxiv.org/abs/2205.04422 ?
___
*[`geometry/optimization/graph_of_convex_sets.cc` line 280 at r4](https://reviewable.io/reviews/robotlocomotion/drake/17397#-NAkTKAb8lJeVA6x-F_t:-NAkTKAb8lJeVA6x-F_u:btf4jbj) ([raw file](https://github.com/robotlocomotion/drake/blob/b21d16d9bf564a6516237fdc31adb65ee6bea639/geometry/optimization/graph_of_convex_sets.cc#L280)):*
> ```cpp
>     incoming_edges[e->v().id()].push_back(edge_count);
> 
>     // Turn off edges into source
> ```

or out of the target...
___
*[`geometry/optimization/graph_of_convex_sets.cc` line 289 at r4](https://reviewable.io/reviews/robotlocomotion/drake/17397#-NAkYLwz2zebO4qreFP1:-NAkYLwz2zebO4qreFP2:b-el3jl9) ([raw file](https://github.com/robotlocomotion/drake/blob/b21d16d9bf564a6516237fdc31adb65ee6bea639/geometry/optimization/graph_of_convex_sets.cc#L289)):*
> ```cpp
>   int nE = edges_.size();
> 
>   MathematicalProgram prog;
> ```

can you write the mathematical program in a comment the way we do elsewhere?  it's hard to parse.
___
*[`geometry/optimization/graph_of_convex_sets.cc` line 346 at r4](https://reviewable.io/reviews/robotlocomotion/drake/17397#-NAkXGwU8A6v-ilrStEq:-NAkXGwU8A6v-ilrStEr:bjwq2z0) ([raw file](https://github.com/robotlocomotion/drake/blob/b21d16d9bf564a6516237fdc31adb65ee6bea639/geometry/optimization/graph_of_convex_sets.cc#L346)):*
> ```cpp
> 
>   for (const auto& [edge_id, e] : edges_) {
>     // Update bounds of consevation of flow.
> ```

typo: conservation
___
*[`geometry/optimization/graph_of_convex_sets.cc` line 371 at r4](https://reviewable.io/reviews/robotlocomotion/drake/17397#-NAkYYiL1OZv08Ampydu:-NAkYYiL1OZv08Ampydv:b-9ihagx) ([raw file](https://github.com/robotlocomotion/drake/blob/b21d16d9bf564a6516237fdc31adb65ee6bea639/geometry/optimization/graph_of_convex_sets.cc#L371)):*
> ```cpp
>     degree.at(e->v().id()).evaluator()->set_bounds(Vector1d(0), Vector1d(0));
> 
>     // Check if edge e = (u,v) is redundant.
> ```

not redundant, but not on the path from start to goal... right?

**Review by mpetersen94 — COMMENTED** [review]
Done! Did both.

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/17397#-:-NAz2lss99NSRnaEmeRw:bm99wxy)* status: 5 unresolved discussions, LGTM missing from assignee RussTedrake(platform), needs at least two assigned reviewers, missing label for release notes (waiting on @RussTedrake)

___
*[`geometry/optimization/graph_of_convex_sets.h` line 37 at r4](https://reviewable.io/reviews/robotlocomotion/drake/17397#-NAkSpKD7tuaf4bVFpkZ:-NAyl20g2tNuW-s6jRTt:bs20eb5) ([raw file](https://github.com/robotlocomotion/drake/blob/b21d16d9bf564a6516237fdc31adb65ee6bea639/geometry/optimization/graph_of_convex_sets.h#L37)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

i would think it should be true by default?
</blockquote></details>

Done. This doubles the runtime of simple problems so I've turned it off for the test cases that are not testing it explicitly but for users on by default makes sense.
___
*[`geometry/optimization/graph_of_convex_sets.h` line 353 at r4](https://reviewable.io/reviews/robotlocomotion/drake/17397#-NAkSRsM5g8oVOnh-QVW:-NAylXbk0xId4iRODZX0:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/b21d16d9bf564a6516237fdc31adb65ee6bea639/geometry/optimization/graph_of_convex_sets.h#L353)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

btw: perhaps "removes all constraints added to any edge with..?"
</blockquote></details>

Done.
___
*[`geometry/optimization/graph_of_convex_sets.cc` line 266 at r4](https://reviewable.io/reviews/robotlocomotion/drake/17397#-NAkY0QkAlARnmMdunpD:-NAylyEd0E99Cey3NwlF:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/b21d16d9bf564a6516237fdc31adb65ee6bea639/geometry/optimization/graph_of_convex_sets.cc#L266)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

can we perhaps cite Appendix A.2 in https://arxiv.org/abs/2205.04422 ?
</blockquote></details>

Done.
___
*[`geometry/optimization/graph_of_convex_sets.cc` line 280 at r4](https://reviewable.io/reviews/robotlocomotion/drake/17397#-NAkTKAb8lJeVA6x-F_t:-NAym1mg1JtsWN95hhVH:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/b21d16d9bf564a6516237fdc31adb65ee6bea639/geometry/optimization/graph_of_convex_sets.cc#L280)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

or out of the target...
</blockquote></details>

Done.
___
*[`geometry/optimization/graph_of_convex_sets.cc` line 289 at r4](https://reviewable.io/reviews/robotlocomotion/drake/17397#-NAkYLwz2zebO4qreFP1:-NAzBEuI3n2iSnEqhe8u:b-qsvnoa) ([raw file](https://github.com/robotlocomotion/drake/blob/b21d16d9bf564a6516237fdc31adb65ee6bea639/geometry/optimization/graph_of_convex_sets.cc#L289)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

can you write the mathematical program in a comment the way we do elsewhere?  it's hard to parse.
</blockquote></details>

Done. PTAL.
___
*[`geometry/optimization/graph_of_convex_sets.cc` line 346 at r4](https://reviewable.io/reviews/robotlocomotion/drake/17397#-NAkXGwU8A6v-ilrStEq:-NAymDlvAKPgUD4RRj21:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/b21d16d9bf564a6516237fdc31adb65ee6bea639/geometry/optimization/graph_of_convex_sets.cc#L346)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

typo: conservation
</blockquote></details>

Done.
___
*[`geometry/optimization/graph_of_convex_sets.cc` line 371 at r4](https://reviewable.io/reviews/robotlocomotion/drake/17397#-NAkYYiL1OZv08Ampydu:-NAymQh_-ZxFZUqcYE2p:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/b21d16d9bf564a6516237fdc31adb65ee6bea639/geometry/optimization/graph_of_convex_sets.cc#L371)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

not redundant, but not on the path from start to goal... right?
</blockquote></details>

Done.

**Review by RussTedrake — COMMENTED** [review]
:lgtm:
+@rpoyner-tri for platform review, please.

> Reviewed 4 of 4 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/17397#-:-NBEW7B3CLe-yxggO1Yh:b-wavrth)* status: LGTM missing from assignee rpoyner-tri(platform), missing label for release notes (waiting on @rpoyner-tri)

___
*[`geometry/optimization/graph_of_convex_sets.cc` line 299 at r5](https://reviewable.io/reviews/robotlocomotion/drake/17397#-NBEVS7e6QGCi97Pp3q3:-NBEVS7e6QGCi97Pp3q4:b-7s3gwp) ([raw file](https://github.com/robotlocomotion/drake/blob/8070d310ef2b9c27d546a0889dedb33a6f642572/geometry/optimization/graph_of_convex_sets.cc#L299)):*
> ```cpp
>   VectorXDecisionVariable f = prog.NewContinuousVariables(nE, "flow_su");
>   Binding<solvers::BoundingBoxConstraint> f_limits =
>       prog.AddBoundingBoxConstraint(VectorXd::Zero(nE), VectorXd::Ones(nE), f);
> ```

fyi -- we have an overload for this method that accepts scalar bounds, too.

**Review by mpetersen94 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/17397)* status: LGTM missing from assignee rpoyner-tri(platform), missing label for release notes (waiting on @rpoyner-tri and @RussTedrake)

___
*[`geometry/optimization/graph_of_convex_sets.cc` line 299 at r5](https://reviewable.io/reviews/robotlocomotion/drake/17397#-NBEVS7e6QGCi97Pp3q3:-NBGbPqgDLRMtabl1b8B:boyexb3) ([raw file](https://github.com/robotlocomotion/drake/blob/8070d310ef2b9c27d546a0889dedb33a6f642572/geometry/optimization/graph_of_convex_sets.cc#L299)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

fyi -- we have an overload for this method that accepts scalar bounds, too.
</blockquote></details>

Done. Thanks!

**Review by rpoyner-tri — COMMENTED** [review]
:lgtm: missing tests for new bindings

> Reviewed 2 of 5 files at r4, 3 of 4 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/17397#-:-NBI_heH0F5I57Hd1NHQ:b-c4u183)* status: 1 unresolved discussion, missing label for release notes (waiting on @mpetersen94)

___
*[`bindings/pydrake/geometry_py_optimization.cc` line 435 at r6](https://reviewable.io/reviews/robotlocomotion/drake/17397#-NBIXi5z4bY6FuIMduVo:-NBIXi5z4bY6FuIMduVp:b2nxk1h) ([raw file](https://github.com/robotlocomotion/drake/blob/da9e8e1051134c848159c50f8e794f3ceac0a5e8/bindings/pydrake/geometry_py_optimization.cc#L435)):*
> ```cpp
>             .def("ClearAllPhiConstraints",
>                 &GraphOfConvexSets::ClearAllPhiConstraints,
>                 cls_doc.ClearAllPhiConstraints.doc)
> ```

The new bindings in this file should have "bindings tests" -- that is, they don't have to exercise all the semantics, but they do have to demonstrate that Python can use them without crashing. If nothing else, the tests will demonstrate that the string names in this file are spelled correctly.

**Review by mpetersen94 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/17397)* status: 1 unresolved discussion, missing label for release notes (waiting on @rpoyner-tri and @RussTedrake)

___
*[`bindings/pydrake/geometry_py_optimization.cc` line 435 at r6](https://reviewable.io/reviews/robotlocomotion/drake/17397#-NBIXi5z4bY6FuIMduVo:-NBIfWjH0y6mAfPC3n9v:b-3ugh27) ([raw file](https://github.com/robotlocomotion/drake/blob/da9e8e1051134c848159c50f8e794f3ceac0a5e8/bindings/pydrake/geometry_py_optimization.cc#L435)):*
<details><summary><i>Previously, rpoyner-tri (Rick Poyner (rico)) wrote…</i></summary><blockquote>

The new bindings in this file should have "bindings tests" -- that is, they don't have to exercise all the semantics, but they do have to demonstrate that Python can use them without crashing. If nothing else, the tests will demonstrate that the string names in this file are spelled correctly.
</blockquote></details>

Done. Not sure how I missed that!

**Review by rpoyner-tri — COMMENTED** [review]
> Reviewed 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/17397)* status: :shipit: complete! all discussions resolved, LGTM from assignees rpoyner-tri(platform),RussTedrake(platform) (waiting on @mpetersen94)

File: geometry/optimization/graph_of_convex_sets.cc [hunk 1]
```diff
@@ -258,12 +263,164 @@ std::string GraphOfConvexSets::GetGraphvizString(
...
+
+    // Degree constraints (redundant if indegree of w is 0):
+    // 0 <= ∑ f_in + ∑ g_in <= 1
+    RowVectorXd A_degree = RowVectorXd::Ones(2 * Ev_in.size());
+    VectorXDecisionVariable fgin(2 * Ev_in.size());
+    for (size_t ii = 0; ii < Ev_in.size(); ++ii) {
+      fgin(ii) = f(Ev_in[ii]);
+      fgin(Ev_in.size() + ii) = g(Ev_in[ii]);
+    }
+    degree.insert({vertex_id, prog.AddLinearConstraint(A_degree, 0, 1, fgin)});
```
Comment (TobiaMarcucci) [inline geometry/optimization/graph_of_convex_sets.cc:349]: I would think that adding an explicit lower bound of zero here should not help, rather slow down? (This lower bound is implied by the non negativity of the flows, right?) I can imagine Gurobi and Mosek being faster if you give an explicit redundant bound on a variable, but I think that adding redundant constraints of linear combinations of decision variables should not help.

File: geometry/optimization/graph_of_convex_sets.h [hunk 2]
```diff
@@ -31,6 +31,11 @@ struct GraphOfConvexSetsOptions {
   relaxation is tight. */
   bool convex_relaxation{true};
 
+  /** Performs a preprocessing step to remove edges that cannot lie on the
+  path from source to target. In most cases, preprocessing causes a net
+  reduction in computation by reducing the size of the optimization solved. */
+  bool preprocessing{true};
```
Comment (TobiaMarcucci) [inline geometry/optimization/graph_of_convex_sets.h:37]: This has been discussed above but, if can say my opinion, I would make the preprocessing inactive by default, since as Mark said in many problems it substantially slow down the solve. I think that for the maze problem this slow down is quite dramatic (~10x), at least it was with the python overhead. I'd imagine a user being more annoyed by realizing that we are running a preprocessing that slows things down, rather than realizing that he/she can decrease the runtimes by turning on an extra feature that we provide?

File: geometry/optimization/graph_of_convex_sets.h [hunk 3]
```diff
@@ -31,6 +31,11 @@ struct GraphOfConvexSetsOptions {
   relaxation is tight. */
   bool convex_relaxation{true};
 
+  /** Performs a preprocessing step to remove edges that cannot lie on the
+  path from source to target. In most cases, preprocessing causes a net
```
Comment (TobiaMarcucci) [inline geometry/optimization/graph_of_convex_sets.h:35]: I would add an extra sentence of the form: "This preprocessing is not exact, i.e., there could be edges that do not lie on any s-t path that this preprocessing does not identify and eliminate."

**Comment by TobiaMarcucci** [comment]
Sorry if I comment this PR only now, I'm traveling with intermittent internet access. I've made just some minor comments. Feel free to neglect them if you don't want to reopen the PR.

**Comment by RussTedrake** [comment]
I don't see any comments from you @tobiamarcucci ?

**Review by TobiaMarcucci — COMMENTED** [review]
Sorry, now they should be visible.
